indexer.ts

```diff
@@ -0,0 +1,743 @@
+        amount: payout["amount"],
+        ft_id: payout["ft_id"] || null,
+        tx_hash: receiptId,
+      };
+      await context.db.PotPayout.insert(potPayout);
+    }
+  }
+
+  async function handlePayout(args, receiverId, signerId, receiptId) {
+    let data = JSON.parse(args);
+    console.log("fulfill payout data::", { ...data }, receiverId);
+    let payout = {
+      recipient_id: data.project_id,
+      amount: data.amount,
+      paid_at: data.paid_at,
+      tx_hash: receiptId,
+    };
+    await context.db.PotPayout.update({ recipient_id: data.project_id }, payout);
+  }
+
+  async function handlePayoutChallenge(args, receiverId, signerId, receiptId) {
+    let data = JSON.parse(args);
+    console.log("challenging payout..: ", { ...data }, receiverId);
+    let created_at = new Date(
+      Number(BigInt(block.header().timestampNanosec) / BigInt(1000000)) // convert to ms then date
+    )
+    let payoutChallenge = {
+      challenger_id: signerId,
+      pot_id: receiverId,
+      created_at,
+      message: data.reason,
+      tx_hash: receiptId,
+    };
+    await context.db.PotPayoutChallenge.insert(payoutChallenge);
+
+    let activity = {
+      signer_id: signerId,
+      receiver_id: receiverId,
+      timestamp: created_at,
+      type: "Challenge_Payout",
+      action_result: receiverId,
+      tx_hash: receiptId,
+    };
+
+    await context.db.Activity.insert(activity);
+  }
+
+  async function handleListAdminRemoval(args, receiverId, signerId, receiptId) {
+    let data = JSON.parse(args);
+    console.log("removing admin...: ", { ...data }, receiverId);
+    let ts = new Date(
+      Number(BigInt(block.header().timestampNanosec) / BigInt(1000000)) // convert to ms then date
+    )
+    let list = await context.db.List.select({id: receiverId})
+
+    for (const acct in data.admins) {
+      await context.db.ListAdmin.delete({ list_id: list[0].id, admin_id: acct})
+    }
+
+    let activity = {
+      signer_id: signerId,
+      receiver_id: receiverId,
+      timestamp: ts,
+      type: "Remove_List_Admin",
+      tx_hash: receiptId,
+    };
+
+    await context.db.Activity.insert(activity);
+  }
+
+  const GECKO_URL = "https://api.coingecko.com/api/v3";
+  function formatDate(date) {
+    const year = date.getFullYear();
+    const month = String(date.getMonth() + 1).padStart(2, "0");
+    const day = String(date.getDate()).padStart(2, "0");
+
+    return `${day}-${month}-${year}`;
+  }
+
+  // helper function to format
+  function formatToNear(yoctoAmount) {
+    const nearAmount = yoctoAmount / 10 ** 24;
+    return nearAmount;
+  }
+
+  async function handleNewDonations(
+    args,
+    receiverId,
+    signerId,
+    actionName,
+    receiptId
+  ) {
+    let matching_pool = JSON.parse(args).matching_pool
+    if ((actionName == "donate" && receiverId != "donate.potlock.near") || matching_pool) {
+      console.log("calling donate on projects...");
+      return;
+    }
```

Comment:
can you give more insight on what is happening here?

Reply:
so the `handleNewDOnations` function is used on both direct and public donations, and both donation action starts with the `donate` function call, but while calling donate directly on the donations contract executes in one function and returns the needed result, calling it on projects makes a promise call to other call back functions which we also listen to, so for this scenario we do not want to go further if the `donate` call was made on a project, because we'll pick up the callback later on.
I basically could have made different functions for the two cases, so i won't need that conditional statement you highlighted.